kame/kame/mping/mprecv.8
.\"	$KAME: mprecv.8,v 1.3 2000/11/10 12:12:11 itojun Exp $
.\"
.\" Copyright (C) 1999 WIDE Project.
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the project nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd Nov 13, 1999
.Dt MPRECV 1
.Os KAME
.\"
.Sh NAME
.Nm mprecv
.Nd a daemon program to receive multicast udp packets
.\"
.Sh SYNOPSIS
.Nm mprecv
.Op Fl d
.Op Fl p Ar port
.Op Fl v
.Op Fl D Ar dumpfile
.Op Fl H Ar history
.Op Fl L Ar logfile
.Op Fl T Ar loginterval
.Op Fl X Ar expire
.Fl i Ar interface
.Fl m Ar mcastaddr
.\"
.Sh DESCRIPTION
The
.Nm
program is designed for measure multicast routing stability in IPv6
network used with
.Xr mpsend 8
program.
The
.Nm
receives UDP packets sent to multicast address
.Ar mcastaddr
using
.Ar interface .
.Pp
The options are as follows:
.Bl -tag -width Fl
.It Fl d
Specify the
.Nm
to run in daemon mode.
.It Fl p Ar port
Specify the UDP port number.
Default value is 9998.
.It Fl D Ar dumpfile
Specify the name of dumpfile.
By default
.Pa /tmp/mprecv.dump
is used.
.It Fl H Ar history
Specify the number of "packet lost" history kept for each logging
interval.
.It Fl L Ar logfile
Specify the name of logfile.
By default
.Pa /var/log/mprecv.log
is used.
.It Fl T Ar loginterval
Specify the interval of status dump to
.Ar logfile
in second.
The default value is 3600 sec (1 hour).
.It Fl X Ar expire
Specify the expire time in day to delete idle session records.
Default value is 7 days.
.It Fl v
Enable the debug output and prevent the
.Nm
go into daemon mode.
.El
.Pp
The
.Nm
program also writes status dump to
.Ar dumpfile
when it receives a SIGINFO signal.
SIGWINCH signal reset the interval measuring variables.
.\"
.Sh SEE ALSO
.Xr mpsend 8
.\"
.Sh HISTORY
The
.Nm
command first appeared in WIDE/KAME IPv6 protocol stack kit.
.\"
.Sh BUGS
IPv4 is not supported.


CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.
	  as there's no official DOI assignment yet, we use private DOI number.

Tue Jul 15 20:02:05 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_rijndael.c: simplify and update rijndael code.
	  markus@openbsd

2003-07-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: revised the relay agent implementation
	based on dhcpv6-28.
	- dhcp6relay now relays packets between clients and servers using
	  relay forward/reply messages.
	- dhcp6s now accepts relay forward messages and replies with relay
	  reply messages.

Sat Jul 12 15:05:23 JST 2003	suz@crl.hitachi.co.jp
	* openbsd/sys/netinet/igmp_var.h
	makes IGMPV3-kernel compilable on openbsd.

Thu Jul 10 21:35:04 JST 2003	keiichi@iij.ad.jp
	* kame/kame/had/halist.c
	fixed a bug that a DHAAD reply message only includes one global
	address per home agent.  the message must include all global
	addresses of all home agents.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

2003/07/10 21:07:50 JST kjc@csl.sony.co.jp
	the first step for transition to pf/altq.
	marge pf/altq into the existing altq-3.
	allow both pf/altq and altq-3 to coexit.
	currently, ALTQ3_COMPAT and ALTQ3_CLFIER_COMPAT are defined
	in sys/altq/altq.h to support altq-3.
	altq-3 will remain in kame for research experiments but will
	not be merged into bsd releases.

2003-07-08  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* kame/sys/netinet6/{in6_ifattach.c,ip6_mroute.c,ip6_mroute.h}
	protect against interface removal.

Tue Jul  8 18:51:06 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c
	insert alternate careof address sub-option when creating a
	binding update for the home registration.  This sub-option is
	MUST when we protect the message with ESP and strictly speaking,

kame/kame/mping/mprecv.c
/*
 * Copyright (C) 1999 WIDE Project.
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the project nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*	$Id: mprecv.c,v 1.3 2000/12/05 08:51:35 itojun Exp $	*/

#include "mping.h"

struct lost_history {
	time_t	lh_time;	/* expected arrival time of first lost packet */
	u_long	lh_duration;	/* duration of lost packets */
	int	lh_reported;	/* if 1, it is already logged */
};

struct session {
	u_long	s_rid;			/* struct index */
	struct	sockaddr_in6 saddr;
	struct	timeval s_interval;
	u_long	s_count;		/* packet count */
	u_long	s_id;			/* session id */
	u_long	ss_init;		/* initial sequence # */
	u_long	ss_last;		/* last sequence # */
	struct	timeval	st_init;	/* initial received */
	struct	timeval	st_last;	/* last received */

	u_long	sl_count;		/* lost count */
	u_long	sl_packet;		/* lost packet count */
	struct	lost_history	s_lh[MAX_LOST_HISTORY];
	int	s_lh_overflow;

	u_long	si_num;			/* number of samples */
	double 	si_sum;			/* sum of receive interval */
	double 	si_sum2;		/* sum of sqr(receive interval) */

	struct session *s_next;
};

struct session *slist;

char *port = DEFAULT_PORT;
char *maddr = NULL;
char *ifname = NULL;
int daemonize = 0;
int verbose = 0;
int interval = DEFAULT_LOG_INTERVAL;
char *logfile = DEFAULT_LOGFILE;
char *dumpfile = DEFAULT_DUMPFILE;
time_t ses_expire = DEFAULT_SESSION_EXPIRE;
FILE *lfp = NULL;
int srid = 1;
char *argv0;
int nhistory = MAX_LOST_HISTORY;

struct mping *mp;
struct addrinfo *res;

static char *month[] = {
	"Jan", "Feb", "Mar", "Apr", "May", "Jun",
	"Jul", "Aug", "Sep", "Oct", "Nov", "Dec",
};

void receive __P((int));
void move_lost_history __P((struct session *));
void log_open __P((char *, char *));
void log_close __P((void));
void log_date_msg __P((char *));
void log_sp __P((struct session *, int));
void log_output __P((int));
void clear_interval __P((int));
void usage __P((void));

int
main(argc, argv)
	int argc;
	char **argv;
{
	int ch, s, error;
	struct addrinfo hints;
	FILE *fp;

	argv0 = *argv;
	while ((ch = getopt(argc, argv, "di:m:p:vD:H:L:T:X:")) != -1)
		switch (ch) {
		case 'd':
			daemonize = 1;
			break;
		case 'i':
			ifname = optarg;
			break;
		case 'm':
			maddr = optarg;
			break;
		case 'p':
			port = optarg;
			break;
		case 'v':
			verbose = 1;
			break;
		case 'D':
			dumpfile = optarg;
			break;
		case 'H':
			nhistory = atoi(optarg);
			if (nhistory > MAX_LOST_HISTORY) {
				fprintf(stderr, "Max # of lost history is %d\n",
					MAX_LOST_HISTORY);
				nhistory = MAX_LOST_HISTORY;
			}
			break;
		case 'L':
			logfile = optarg;
			break;
		case 'T':
			interval = atoi(optarg);
			break;
		case 'X':
			ses_expire = atoi(optarg) * 3600 * 24;
			break;
		default:
			usage();
			exit(0);
		}
	argc -= optind;
	argv += optind;

	if (daemonize && getuid() != 0) {
		fprintf(stderr, "Only superuser can specify daemon mode\n");
		exit(-1);
	}

	memset(&hints, 0, sizeof(hints));
	hints.ai_family = PF_INET6;
	hints.ai_socktype = SOCK_DGRAM;
	hints.ai_flags = AI_PASSIVE;
	error = getaddrinfo(maddr, port, &hints, &res);
	if (error)
		err(1, "%s", gai_strerror(error));

	if (verbose) {
		int i;
		u_char *p;

		fprintf(stderr, "%s configuration is:\n", argv0);
		fprintf(stderr, "\tport:        %s\n", port);
		fprintf(stderr, "\tai_family:   %d\n", res->ai_family);
		fprintf(stderr, "\tai_socktype: %d\n", res->ai_socktype);
		fprintf(stderr, "\tai_protocol: %d\n", res->ai_protocol);
		fprintf(stderr, "\tai_addrlen:  %d\n", res->ai_addrlen);
		fprintf(stderr, "\tai_addr:     ");
		for (i = res->ai_addrlen, p = (u_char *)res->ai_addr;
			i; i--, p++) {
			fprintf(stderr, "%02x ", *p);
			if (i == 21)
				fprintf(stderr, "\n\t\t     ");
		}
		fprintf(stderr, "\n");
	}

	s = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
	if (s < 0)
		err(1, NULL);
	if (bind(s, res->ai_addr, res->ai_addrlen) < 0)
		err(1, NULL);

	if (IN6_IS_ADDR_MULTICAST(&(((struct sockaddr_in6 *)(res->ai_addr))->sin6_addr))) {
		struct ipv6_mreq mreq6;

		if (ifname == NULL) {
			fprintf(stderr, "ifname is required for multicast\n");
			exit(-1);
		}

		if ((mreq6.ipv6mr_interface = if_nametoindex(ifname)) == 0)
			err(1, "if_nametoindex");
		if (verbose)
			fprintf(stderr, "\tifindex: %d\n",
				mreq6.ipv6mr_interface);
		memcpy(&mreq6.ipv6mr_multiaddr,
			&((struct sockaddr_in6 *)res->ai_addr)->sin6_addr,
			sizeof(mreq6.ipv6mr_multiaddr));
		if (setsockopt(s, IPPROTO_IPV6, IPV6_JOIN_GROUP,
			&mreq6, sizeof(mreq6)))
			err(1, "setsockopt(IPV6_JOIN_GROUP)");
	}

	if (daemonize && verbose == 0) {
		if (daemon(0, 0) != 0)
			err(1, "daemon");
		if ((fp = fopen(PID_FILE, "w")) == NULL)
			err(1, PID_FILE);
		fprintf(fp, "%d\n", getpid());
		fclose(fp);
	} else {
		lfp = stdout;
	}

	log_open(logfile, "a");
	log_date_msg("mprecv start");
	log_close();

	signal(SIGINT, log_output);
	signal(SIGTERM, log_output);
	signal(SIGQUIT, log_output);
	signal(SIGINFO, log_output);
	signal(SIGALRM, log_output);
	signal(SIGWINCH, clear_interval);

	alarm(interval);

	for (;;) {
		receive(s);
	}

	exit(0);
	/*NOTREACHED*/
}

void
receive(s)
	int s;
{
	int cc, fromlen;
	double t_int;
	struct timeval tv;
	struct sockaddr_in6 from6;
	struct session *sp;
	struct mping *mp;
	u_char *buf[MAX_MSGSIZE];

	fromlen = sizeof(from6);
	cc = recvfrom(s, buf, sizeof(buf), 0,
			(struct sockaddr *)&from6, &fromlen);
	if (cc < 0)
		err(1, "recvfrom");
	if (verbose)
		printf("%d bytes recvd\n", cc);
	mp = (struct mping *)buf;
	for (sp = slist; sp; sp = sp->s_next) {
		if (IN6_ARE_ADDR_EQUAL(&from6.sin6_addr,
			&sp->saddr.sin6_addr))
			break;
	}
	gettimeofday(&tv, NULL);
	if (sp == NULL)	{			/* allocate new session */
		sp = (struct session *)malloc(sizeof(struct session));
		memset(sp, 0, sizeof(struct session));
		sp->s_next = slist;
		slist = sp;
		memcpy(&sp->saddr, &from6, sizeof(struct sockaddr_in6));
		sp->st_init = tv;
		sp->ss_init = mp->m_seq;
		sp->s_rid = srid++;
		sp->s_id = mp->m_sessid;
		sp->ss_last = mp->m_seq;
		log_open(logfile, "a");
		log_date_msg("new sender detected");
		log_sp(sp, 0);
		log_close();
	} else if (sp->s_id != mp->m_sessid) {	/* sender restarted */
		/* dump info regarding the last session */
		log_open(logfile, "a");
		log_date_msg("session terminiation detected");
		log_sp(sp, 0);
		log_close();
		sp->st_init = tv;
		sp->ss_init = mp->m_seq;
		sp->s_id = mp->m_sessid;
		sp->s_count = 0;
		sp->sl_count = 0;
		sp->sl_packet = 0;
		move_lost_history(sp);
		sp->s_lh[0].lh_time = 0;
		sp->s_lh[0].lh_duration = 0;
		sp->s_lh[0].lh_reported = 0;
	} else if (sp->ss_last > 0 && sp->ss_last + 1 < mp->m_seq) {
		move_lost_history(sp);
		sp->s_lh[0].lh_time =
			sp->st_last.tv_sec + mp->m_interval.tv_sec;
		sp->s_lh[0].lh_duration = tv.tv_sec - sp->s_lh[0].lh_time;
		sp->s_lh[0].lh_reported = 0;
		sp->sl_count++;
		sp->sl_packet += mp->m_seq - (sp->ss_last + 1);
	} else if (sp->ss_last > 0) {
		sp->si_num++;
		t_int = tv.tv_sec - sp->st_last.tv_sec +
			(tv.tv_usec - sp->st_last.tv_usec) * 0.000001;
		sp->si_sum += t_int;
		sp->si_sum2 += t_int * t_int;
	}
	sp->s_interval = mp->m_interval;
	sp->st_last = tv;
	sp->ss_last = mp->m_seq;
	sp->s_count++;
}

void
move_lost_history(sp)
	struct session *sp;
{
	int i;

	if (sp->s_lh[nhistory - 1].lh_time > 0)
		sp->s_lh_overflow++;
	for (i = nhistory - 2; i > 0; i--) {
		sp->s_lh[i + 1] = sp->s_lh[i];
	}
	sp->s_lh[0].lh_time = 0;
	sp->s_lh[0].lh_duration = 0;
}

void
log_open(fname, mode)
	char *fname, *mode;
{
	if (lfp != stdout) {
		if ((lfp = fopen(fname, mode)) == NULL)
			err(1, "%s", fname);
	}
}

void
log_close()
{
	if (lfp != stdout) {
		fclose(lfp);
		lfp = NULL;
	}
}

/*
 * Print heading line of each log chunk
 */
void
log_date_msg(msg)
	char *msg;
{
	time_t t;
	struct tm *tm;

	t = time(0);
	tm = localtime(&t);
	fprintf(lfp, "%s %2d %02d:%02d:%02d ========== %s\n",
		month[tm->tm_mon],
		tm->tm_mday, tm->tm_hour, tm->tm_min, tm->tm_sec, msg);
}

/*
 * Write log information to specified session
 */
void
log_sp(sp, force)
	struct session *sp;
	int force;
{
	int i;
	double mean, mean2, sd;
	time_t t;
	struct tm *tm;
	struct lost_history *lhp;
	char buf[BUFSIZ];

	getnameinfo((struct sockaddr *)&sp->saddr, sizeof(sp->saddr),
		buf, sizeof(buf), NULL, 0, NI_NUMERICHOST);
	fprintf(lfp, "[%02ld] host %s (%lu)\n", sp->s_rid,
		buf, sp->s_id);
	t = (time_t)sp->st_last.tv_sec;
	tm = localtime(&t);
	fprintf(lfp, "[%02ld] last read %s %2d %02d:%02d:%02d,",
		sp->s_rid, month[tm->tm_mon],
		tm->tm_mday, tm->tm_hour, tm->tm_min, tm->tm_sec);
	fprintf(lfp, " elapsed %lu sec, last read %lu sec ago\n",
		sp->st_last.tv_sec - sp->st_init.tv_sec,
		time(0) - sp->st_last.tv_sec);
	fprintf(lfp, "[%02ld] seq init %lu last %lu diff %lu\n", sp->s_rid,
		sp->ss_init, sp->ss_last,
		sp->ss_last - sp->ss_init);
	if (sp->si_num > 0) {
		mean = sp->si_sum / sp->si_num;
		mean2 =  sp->si_sum2 / sp->si_num;
		sd = sqrt(mean2 - mean * mean);
		fprintf(lfp, "[%02ld] %ld sample mean %6.3f sec sd %6.3f sec\n",
			sp->s_rid, sp->si_num, mean, sd);
	}
	fprintf(lfp, "[%02ld] count total %ld lost %ld lost packets %ld\n",
		sp->s_rid, sp->s_count, sp->sl_count, sp->sl_packet);
	for (i = nhistory - 1; i >= 0; i--) {
		lhp = &sp->s_lh[i];
		if (lhp->lh_reported != 0 && force == 0)
			continue;
		if (lhp->lh_time > 0) {
			tm = localtime(&lhp->lh_time);
			fprintf(lfp, "[%02ld] last lost %s %2d %02d:%02d:%02d,",
				sp->s_rid, month[tm->tm_mon],
				tm->tm_mday, tm->tm_hour, tm->tm_min,
				tm->tm_sec);
			fprintf(lfp, " for %ld sec\n", lhp->lh_duration);
			lhp->lh_reported = 1;
		}
	}
	if (sp->s_lh_overflow > 0)
		fprintf(lfp, "[%02ld] %d lost history overwritten\n",
			sp->s_rid, sp->s_lh_overflow);
	if (force == 0)
		sp->s_lh_overflow = 0;
}

void
log_output(sig)
	int sig;
{
	struct session *sp;
	int force;

	force = 0;
	switch (sig) {
	case SIGALRM:
		log_open(logfile, "a");
		log_date_msg("periodic log dump");
		break;
	case SIGINT:
	case SIGTERM:
	case SIGQUIT:
		log_open(logfile, "a");
		log_date_msg("mprecv terminating");
		break;
	case SIGINFO:
		log_open(dumpfile, "w");
		log_date_msg("operator driven log dump");
		force = 1;
		break;
	}

	for (sp = slist; sp; sp = sp->s_next)
		log_sp(sp, force);

	log_close();

	if (sig == SIGALRM)
		alarm(interval);
	if (sig == SIGINT || sig == SIGTERM || sig == SIGQUIT) {
		fclose(lfp);
		unlink(PID_FILE);
		exit(0);
	}
}

void
clear_interval(sig)
	int sig;
{
	struct session *sp;

	for (sp = slist; sp; sp = sp->s_next) {
		sp->si_num = 0;
		sp->si_sum = 0.0;
		sp->si_sum2 = 0.0;
	}
}

void
usage()
{
	fprintf(stderr,
"Usage: %s [-dv] [-p port] [-D dumpfile] [-H history] [-i interface] \n"
"        [-L logfile] [-m mcastaddr] [-T loginterval] [-X expire]\n", argv0);
}
